File: python/core/groupby.py
# -*- coding: utf-8 -*-

from __future__ import absolute_import, print_function, division, unicode_literals

import random

from pyspark import SparkConf, SparkContext


if __name__ == '__main__':
    conf = SparkConf().setAppName('GroupBy')
    sc = SparkContext(conf=conf)

    nums = sc.parallelize(range(0, 5), 5).flatMap(
        lambda i: [random.randint(0, 10) for _ in range(0, 1000)]
    ).persist()

    print('### Count ###')
    print(nums.count())
    print()

    print('### Average: reduce ###')
    avg1 = nums.map(lambda n: (n, 1)).reduce(lambda x, y: (x[0] + y[0], x[1] + y[1]))
    print(avg1[0] / avg1[1])
    print()

    print('### Average: fold ###')
    avg2 = nums.map(lambda n: (n, 1)).fold((0, 0), lambda x, y: (x[0] + y[0], x[1] + y[1]))
    print(avg2[0] / avg2[1])
    print()

    print('### Average: aggregate ###')
    avg3 = nums.aggregate(
        (0, 0),
        lambda acc, v: (acc[0] + v, acc[1] + 1),
        lambda acc1, acc2: (acc1[0] + acc2[0], acc1[1] + acc2[1])
    )
    print(avg3[0] / avg3[1])
    print()

    print('### Grouping: reduceByKey ###')
    grouped = nums.map(lambda n: (n, 1)).reduceByKey(lambda x, y: x + y).sortByKey().collect()
    for (n, c) in grouped:
        print(n, c)
    print()

    print('### Grouping: foldByKey ###')
    grouped = nums.map(lambda n: (n, 1)).foldByKey(0, lambda x, y: x + y).sortByKey().collect()
    for (n, c) in grouped:
        print(n, c)
    print()

    print('### Grouping: aggregateByKey ###')
    grouped = nums.map(lambda n: (n, 1)).aggregateByKey(
        0,
        lambda acc, v: acc + v,
        lambda acc1, acc2: acc1 + acc2
    ).sortByKey().collect()
    for (n, c) in grouped:
        print(n, c)
    print()

    print('### Grouping: combineByKey ###')
    grouped = nums.map(lambda n: (n, 1)).combineByKey(
        lambda v: v,
        lambda acc, v: acc + v,
        lambda acc1, acc2: acc1 + acc2
    ).sortByKey().collect()
    for (n, c) in grouped:
        print(n, c)
    print()

    sc.stop()

File: scala/src/main/scala/core/GroupBy.scala
package core

import scala.collection.mutable.ListBuffer
import scala.util.Random

import org.apache.spark.SparkContext._
import org.apache.spark.{SparkConf, SparkContext}

object GroupBy {

  def main(args: Array[String]) {
    run()
  }

  def run() {
    val conf = new SparkConf().setAppName("GroupBy")
    val sc = new SparkContext(conf)

    val nums = sc.parallelize(0 until 5, 5).flatMap { _ =>
      val rnd = new Random
      val buffer = ListBuffer[Int]()
      for (i <- 0 until 1000) {
        buffer += rnd.nextInt(10)
      }
      buffer.toList
    }.persist()

    println("### Count ###")
    println(nums.count())
    println()

    println("### Average: reduce ###")
    val avg1 = nums.map((_, 1)).reduce { case (x, y) => (x._1 + y._1, x._2 + y._2) }
    println(avg1._1 * 1.0 / avg1._2)
    println()

    println("### Average: fold ###")
    val avg2 = nums.map((_, 1)).fold((0, 0)) { case (x, y) => (x._1 + y._1, x._2 + y._2) }
    println(avg2._1 * 1.0 / avg2._2)
    println()

    println("### Average: aggregate ###")
    val avg3 = nums.aggregate((0, 0))(
      (acc, v) => (acc._1 + v, acc._2 + 1),
      (acc1, acc2) => (acc1._1 + acc2._1, acc1._2 + acc2._2)
    )
    println(avg3._1 * 1.0 / avg3._2)
    println()

    println("### Grouping: reduceByKey ###")
    nums.map((_, 1)).reduceByKey(_ + _).sortByKey().collect().foreach(println)
    println()

    println("### Grouping: foldByKey ###")
    nums.map((_, 1)).foldByKey(0)(_ + _).sortByKey().collect().foreach(println)
    println()

    println("### Grouping: aggregateByKey ###")
    nums.map((_, 1)).aggregateByKey(0)(
      (acc, v) => acc + v,
      (acc1, acc2) => acc1 + acc2
    ).sortByKey().collect().foreach(println)
    println()

    println("### Grouping: combineByKey ###")
    nums.map((_, 1)).combineByKey(
      (v: Int) => v,
      (acc: Int, v: Int) => acc + v,
      (acc1: Int, acc2: Int) => acc1 + acc2
    ).sortByKey().collect().foreach(println)
    println()

    sc.stop()
  }
}
